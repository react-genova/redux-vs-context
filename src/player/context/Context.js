import { createContext } from 'react';
import { createSelector } from 'reselect';

const PlayerContext = createContext({
    name: '',
    surname: '',
    onNameChanged: () => null,
    onSurnameChanged: () => null
});

const getPlayer = value => value;
export const getPlayerName = createSelector(
    getPlayer,
    player => player.name
);
export const getPlayerSurname = createSelector(
    getPlayer,
    player => player.surname
);
export const getPlayerNameChangedCallback = createSelector(
    getPlayer,
    player => player.onNameChanged
);
export const getPlayerSurnameChangedCallback = createSelector(
    getPlayer,
    player => player.onSurnameChanged
);

export default PlayerContext;
