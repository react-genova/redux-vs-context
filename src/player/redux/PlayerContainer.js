import { connect } from 'react-redux';
import PlayerEdit from '../components/PlayerEdit';
import { getPlayerName, getPlayerSurname, editPlayerName, editPlayerSurname } from './player';

const mapStateToProps = state => ({
    name: getPlayerName(state),
    surname: getPlayerSurname(state),
});

const mapDispatchToProps = {
    onNameChanged: editPlayerName,
    onSurnameChanged: editPlayerSurname,
};

const PlayerContainer = connect(mapStateToProps, mapDispatchToProps)(PlayerEdit);

export default PlayerContainer;
