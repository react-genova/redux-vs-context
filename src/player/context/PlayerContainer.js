import React from 'react';
import PlayerEdit from '../components/PlayerEdit';
import Context, { getPlayerName, getPlayerNameChangedCallback, getPlayerSurname, getPlayerSurnameChangedCallback } from './Context';

const SurnameContainer = () => (
    <Context.Consumer>
    {player => <PlayerEdit
        surname={getPlayerSurname(player)}
        name={getPlayerName(player)}
        onNameChanged={getPlayerNameChangedCallback(player)}
        onSurnameChanged={getPlayerSurnameChangedCallback(player)} />}
    </Context.Consumer>
);

export default SurnameContainer;
