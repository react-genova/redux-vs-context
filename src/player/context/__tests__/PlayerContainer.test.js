import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import PlayerContainer from '../PlayerContainer';
import Context from '../Context';

jest.mock('../../components/PlayerEdit', () => ({
    __esModule: true,
    default: ({ name, surname, onNameChanged, onSurnameChanged }) => (
        <div>
            <div data-testid="name" onClick={() => onNameChanged(name)}>{name}</div>
            <div data-testid="surname" onClick={() => onSurnameChanged(surname)}>{surname}</div>
        </div>
    )
}));

describe('PlayerContainer consumer container', () => {
    it('renders without crashing and passes values down to its child', () => {
        const onNameChanged = jest.fn();
        const onSurnameChanged = jest.fn();
        const { getByTestId, container } = render((
            <Context.Provider
                value={{
                    name: 'Dino',
                    surname: 'Zoff',
                    onNameChanged,
                    onSurnameChanged,
                }}
            >
                <PlayerContainer />
            </Context.Provider>
        ));
        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId('name').textContent).toBe('Dino');
        expect(getByTestId('surname').textContent).toBe('Zoff');
        fireEvent.click(getByTestId('name'));
        fireEvent.click(getByTestId('surname'));
        expect(onNameChanged).toHaveBeenCalledTimes(1);
        expect(onNameChanged).toHaveBeenCalledWith('Dino');
        expect(onSurnameChanged).toHaveBeenCalledTimes(1);
        expect(onSurnameChanged).toHaveBeenCalledWith('Zoff');
    });
});
