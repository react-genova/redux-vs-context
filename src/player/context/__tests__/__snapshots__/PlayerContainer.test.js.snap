// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerContainer consumer container renders without crashing and passes values down to its child 1`] = `
<div>
  <div
    data-testid="name"
  >
    Dino
  </div>
  <div
    data-testid="surname"
  >
    Zoff
  </div>
</div>
`;
